### PR TITLE
dockershim: use the new Metadata types and add more unit tests

### PR DESCRIPTION
**Pull request by yujuhong**
Part of #31459

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31573)

### CODE REVIEW

File: pkg/kubelet/dockershim/convert.go [hunk 1]
```diff
@@ -49,20 +49,20 @@ func toRuntimeAPIImage(image *dockertypes.Image) (*runtimeApi.Image, error) {
...
-		Metadata: &runtimeApi.ContainerMetadata{
-			Name:    &containerName,
-			Attempt: &attempt,
-		},
+		Id:       &c.ID,
+		Metadata: metadata,
 		Image:    &runtimeApi.ImageSpec{Image: &c.Image},
 		ImageRef: &c.ImageID,
 		State:    &state,
 		Labels:   c.Labels,
```
Comment (feiskyer) [inline pkg/kubelet/dockershim/convert.go]: nit: add a todo for extracting annotations from labels.

File: pkg/kubelet/dockershim/docker_sandbox.go [hunk 2]
```diff
@@ -182,19 +178,24 @@ func (ds *dockerService) ListPodSandbox(filter *runtimeApi.PodSandboxFilter) ([]
...
-			}
-		}
+	for i := range containers {
+		c := containers[i]
 
-		s := toRuntimeAPISandbox(&c)
-		if filterOutReadySandboxes && s.GetState() == runtimeApi.PodSandBoxState_READY {
+		converted, err := toRuntimeAPISandbox(&c)
+		if err != nil {
+			glog.V(5).Infof("Unable to convert docker to runtime API container: %v", err)
```
Comment (feiskyer) [inline pkg/kubelet/dockershim/docker_sandbox.go]: nit: replace `container` with `sandbox`.

File: pkg/kubelet/dockershim/naming.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+	return uint32(n), nil
+}
+
+// TODO: Evaluate whether we should rely on labels completely.
+func parseSandboxName(name string) (*runtimeApi.PodSandboxMetadata, error) {
+	// Docker adds a "/" prefix to names. so trim it.
+	name = strings.TrimPrefix(name, "/")
+
+	parts := strings.Split(name, nameDelimiter)
+	if len(parts) != 6 {
```
Comment (feiskyer) [inline pkg/kubelet/dockershim/naming.go]: nit: also checks whether the first part is equal to kubePrefix here.

File: pkg/kubelet/dockershim/naming.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+	}, nil
+}
+
+// TODO: Evaluate whether we should rely on labels completely.
+func parseContainerName(name string) (*runtimeApi.ContainerMetadata, error) {
+	// Docker adds a "/" prefix to names. so trim it.
+	name = strings.TrimPrefix(name, "/")
+
+	parts := strings.Split(name, nameDelimiter)
+	if len(parts) != 6 {
```
Comment (feiskyer) [inline pkg/kubelet/dockershim/naming.go]: nit: also checks whether the first part is equal to kubePrefix here.

**Comment by yujuhong** [comment]
Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/kubelet/dockershim/convert.go, line 64 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31573#-KQ8klRHHEEL1fOFjSa1-r3-64:-KQMC6Bz8FIYSgCqQZqW:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/9465044db04b52d9bbdaf7c5f16da7190a69420a/pkg/kubelet/dockershim/convert.go#L64)):*
<details><summary><i>Previously, feiskyer (Pengfei Ni) wrote…</i></summary>
> nit: add a todo for extracting annotations from labels.
</details>
Done.

---

*[pkg/kubelet/dockershim/docker_sandbox.go, line 186 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31573#-KQ8klRFyj9anfiLuoAm-r3-186:-KQMCHPX-L6zsrPHgy8o:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/9465044db04b52d9bbdaf7c5f16da7190a69420a/pkg/kubelet/dockershim/docker_sandbox.go#L186)):*
<details><summary><i>Previously, feiskyer (Pengfei Ni) wrote…</i></summary>
> nit: replace `container` with `sandbox`.
</details>
Done.

---

*[pkg/kubelet/dockershim/naming.go, line 89 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31573#-KQ8klRGzXA_jJc4JNH--r3-89:-KQMG8wXx6MEYy8DiE5b:bn21kf1) ([raw file](https://github.com/kubernetes/kubernetes/blob/9465044db04b52d9bbdaf7c5f16da7190a69420a/pkg/kubelet/dockershim/naming.go#L89)):*
<details><summary><i>Previously, feiskyer (Pengfei Ni) wrote…</i></summary>
> nit: also checks whether the first part is equal to kubePrefix here. 
</details>
Done. Also added a test.

---

*[pkg/kubelet/dockershim/naming.go, line 112 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31573#-KQ8klRGzXA_jJc4JNH--r3-112:-KQMGBt_9XWYfAHR4tOn:bn21kf1) ([raw file](https://github.com/kubernetes/kubernetes/blob/9465044db04b52d9bbdaf7c5f16da7190a69420a/pkg/kubelet/dockershim/naming.go#L112)):*
<details><summary><i>Previously, feiskyer (Pengfei Ni) wrote…</i></summary>
> nit: also checks whether the first part is equal to kubePrefix here. 
</details>
Done. Also added a test.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31573)*

**Comment by yujuhong** [comment]
Duh...I forgot to include a unit test file (`naming_test.go`). PTAL again. Thanks!

**Comment by feiskyer** [comment]
LGTM.

**Comment by euank** [comment]
LGTM as well

**Comment by yujuhong** [comment]
Rebased and re-applied lgtm

**Pull request merged**